cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java

```diff
@@ -169,30 +166,20 @@ private void closeDataset(DatasetCacheKey key, Dataset dataset) {
 
     // any transaction aware that is not in the active tx-awares is added to the current tx context (if there is one).
     if (!bypass && dataset instanceof TransactionAware) {
-      WeakReferenceDelegatingTransactionAware txAware =
-        new WeakReferenceDelegatingTransactionAware((TransactionAware) dataset);
-      WeakReferenceDelegatingTransactionAware existing = activeTxAwares.get(key);
+      TransactionAware txAware = (TransactionAware) dataset;
+      TransactionAware existing = activeTxAwares.get(key);
       if (existing == null) {
         activeTxAwares.put(key, txAware);
         if (txContext != null) {
           txContext.addTransactionAware(txAware);
         }
       } else {
-        TransactionAware actual = existing.get();
-        if (actual == null) {
-          activeTxAwares.put(key, txAware);
-          if (txContext != null) {
-            txContext.removeTransactionAware(existing);
-            txContext.addTransactionAware(txAware);
-          }
-        } else {
-          // this better be the same dataset, otherwise the cache did not work
-          if (actual != dataset) {
-            throw new IllegalStateException(
-              String.format("Unexpected state: Cache returned %s for %s, which is different from the " +
-                              "active transaction aware %s for the same key. This should never happen.",
-                            dataset, key, actual));
-          }
+        // this better be the same dataset, otherwise the cache did not work
+        if (existing != dataset) {
```

Comment:
The else and if scopes can be combine to one `else-if` to make the code less clutter.